File: services/server/src/safe_socket/safe_socket.py
import socket


def recv_all(socket: socket.socket, size):
    data = bytearray()
    while len(data) < size:
        chunk = socket.recv(size - len(data))
        if not chunk:
            continue
        data.extend(chunk)
    return bytes(data)


def send_all(socket: socket.socket, bytes):
    total_sent = 0
    while total_sent < len(bytes):
        sent = socket.send(bytes[total_sent:])
        if sent == 0:
            continue
        total_sent += sent
    return total_sent

File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import "io"

func SendAll(socket io.Writer, bytes []byte) error {
	totalSent := 0
	for totalSent < len(bytes) {
		n, err := socket.Write(bytes[totalSent:])
		if err != nil {
			return err
		}
		if n == 0 {
			continue
		}
		totalSent += n
	}
	return nil
}

func RecvAll(socket io.Reader, size int) ([]byte, error) {
	buff := make([]byte, size)
	totalRead := 0
	for totalRead < size {
		n, err := socket.Read(buff[totalRead:])
		if err != nil {
			if err == io.EOF && totalRead < size {
				return nil, io.ErrUnexpectedEOF
			}
			return nil, err
		}
		if n == 0 {
			continue
		}
		totalRead += n
	}
	return buff, nil
}
